File: Dynamic_Programming/MaxSubarraySum/Kadane_Algorithm.java
package Dynamic_Programming.MaxSubarraySum;
/**
 * Write an efficient program to find the sum of contiguous subarray within a one-dimensional array of numbers which has the largest sum.
 * Created by lmsj0 on 9/24/2017.
 */
public class Kadane_Algorithm {
    public static int kadane_algorithm(int []a){
        int y = Integer.MIN_VALUE;
        int x = 0;
        int start = 0;
        int end = 0;
        int k = 0;

        for(int i = 0; i< a.length; i++){
            x = x + a[i];
            if(y < x){
                y = x;
                start = k;
                end = i;
            }
            if(x < 0){
                x = 0;
                k = i + 1;
            }
        }
        System.out.println("start at " + start);
        System.out.println("end at " + end);
        return  y;
    }
    public static void main(String [] args){
        //Scanner input = new Scanner(System.in);
        int [] a = {-22,-31,-4,-16,-2,-91,-5,-3};
        System.out.print("Max contiguous sum is " + kadane_algorithm(a));
    }
}

File: algorithm_solution_By_Python/MaxSubarraySum/Kadane_Algorithm.py
#
#Write an efficient program to find the sum of contiguous subarray within a one-dimensional array of numbers which has the largest sum.
def kadane(a):

    max_so_far =  -10000000
    max_ending = 0
    start = 0
    end = 0
    s = 0

    for i in  range(0,len(a)):
        max_ending = max_ending + a[i]
        if(max_so_far < max_ending):
            max_so_far = max_ending
            start = s
            end = i

        if max_ending < 0:
            max_ending = 0
            s = i + 1

    print( "start at ", start, "end at ", end)


    return max_so_far

a = [-13,-3,-25,-20,-3,-16,-23,-12,-5,-22,-15,-4,-7]
print ("Maximum contiguous sum is " , kadane(a))
